tracer/security_aware_span_listener.go

```diff
@@ -0,0 +1,138 @@
+package tracer
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/thundra-io/thundra-lambda-agent-go/constants"
+	"github.com/thundra-io/thundra-lambda-agent-go/utils"
+	logger "log"
+)
+
+var defaultSecurityMessage = "Operation was blocked due to security configuration"
+
+type SecurityAwareSpanListener struct {
+	block     bool
+	whitelist []Operation
+	blacklist []Operation
+}
+
+func (s *SecurityAwareSpanListener) OnSpanStarted(span *spanImpl) {
+	if !s.isExternalOperation(span) {
+		return
+	}
+
+	for _, op := range s.blacklist {
+		if op.matches(span) {
+			s.handleSecurityIssue(span)
+			return
+		}
+	}
+
+	if len(s.whitelist) > 0 {
+		for _, op := range s.whitelist {
+			if op.matches(span) {
+				return
+			}
+		}
+		s.handleSecurityIssue(span)
+	}
+
+}
+
+func (s *SecurityAwareSpanListener) OnSpanFinished(span *spanImpl) {
+	return
+}
+
+func (s *SecurityAwareSpanListener) PanicOnError() bool {
+	return true
+}
+
+func (s *SecurityAwareSpanListener) handleSecurityIssue(span *spanImpl) {
+	if s.block {
+		err := errors.New(defaultSecurityMessage)
+		span.SetTag(constants.SecurityTags["BLOCKED"], true)
+		span.SetTag(constants.SecurityTags["VIOLATED"], true)
+		utils.SetSpanError(span, err)
+		panic(err)
```

Comment:
can panic'ing be avoided here?

in Go, errors are usually carried as values and panicing considered a bit dangerous, needed to be used with care.

if panicing in this logic is still needed, having them thrown by the caller is a bit nicer otherwise its presence can be forgotten in the implementation details